### PR TITLE
Partial display entity support

### PR DESCRIPTION
**Pull request by tal5**
## Additions

- `PaperAPITools#getText` - to get a text display entity's text.
- `PaperAPITools#setText` - to set a text display entity's text.
- `EntityBrightness`, `EntityDefaultBackground`, `EntityDisplay` (item transform, text alignment), `EntityGlowColor`, `EntityHeight` (culling height), `EntityInterpolationDuration`, `EntityInterpolationStart`, `EntityLineWidth`, `EntityOpacity`, `EntityPivot` (billboard), `EntityScale`, `EntitySeeThrough` (seeing text displays through blocks), `EntityShadowRadius`, `EntityShadowStrength`, `EntityText`, `EntityTextShadowed`, `EntityTranslation`, `EntityViewRange`, `EntityWidth` (culling width).

## Changes
- Fixed come minor alphabetical sorting issues in `PropertyRegistry`.
- `EntityDisplayEntityData` - updated to modern `EntityProperty` format.

## Notes

- `EntityOpacity` is represented by a (signed) byte internally, but is actually meant to be an unsigned `0-255`, so did some conversions there; seems to all work as expected other then some weirdness where `0-3` are completely opaque, although the byte being set looks fine, and it works as expected from `4+`, so I assume it's an unrelated issue?
- `EntityWidth/Height` don't have `culling` in their name as they could be used for interaction entities later on.
- `EntityBrightness` currently stops if you input an invalid block/sky brightness, let me know if that should be changed.
- The `PaperAPITools#getText` Spigot impl converts `null` to an empty string because that's what Paper's method does (which seems to better match what Mojang does as well), makes more sense to have it consistent between both impls.
With some quick testing, I couldn't even get it to return null; Spigot seems to convert a `null` component to an empty string, and a `null` string to an empty component.
- Depends on https://github.com/DenizenScript/Denizen-Core/pull/92 for the ones with nullable mech input.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityInterpolationStart.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+public class EntityInterpolationStart extends EntityProperty<DurationTag> {
+
+    // <--[property]
+    // @object EntityTag
+    // @name interpolation_start
+    // @input DurationTag
+    // @description
+    // The delay between a display entity receiving an update to an interpolated value(s) to it starting its interpolation.
+    // Interpolation is started whenever this value is set. If no changes were made to the entity in the same tick, it will (visually) redo it's last interpolation.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityInterpolationStart.java]: `its`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityPivot.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.bukkit.entity.Display;
+
+public class EntityPivot extends EntityProperty<ElementTag> {
+
+    // <--[property]
+    // @object EntityTag
+    // @name pivot
+    // @input ElementTag
+    // @description
+    // A display entity's pivot point/axes, can be any of <@link url https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/Display.Billboard.html>.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityPivot.java:16]: should use `@synonyms EntityTag.billboard`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityDisplay.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+public class EntityDisplay extends EntityProperty<ElementTag> {
+
+    // <--[property]
+    // @object EntityTag
+    // @name display
+    // @input ElementTag
+    // @description
+    // For an item display entity this is the model transform it will display, can be any of <@link url https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/ItemDisplay.ItemDisplayTransform.html>.
+    // For a text display entity this is its text alignment, can be any of <@link url https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/TextDisplay.html>.
+    // -->
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityDisplay.java:19]: should use `@synonyms EntityTag.display_transform, EntityTag.text_alignment`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityInterpolationDuration.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import org.bukkit.entity.Display;
+
+public class EntityInterpolationDuration extends EntityProperty<DurationTag> {
+
+    // <--[property]
+    // @object EntityTag
+    // @name interpolation_duration
+    // @input DurationTag
+    // @description
+    // The duration a display entity will spend interpolating between interpolated value(s).
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityInterpolationDuration.java:15]: The interpolation things should have `<@link`s between them, and the "can be interpolated" meta should `<@link` somewhere to explain it. Maybe a short language meta as a centerpoint? Either for interpolation specifically, or as a general directory of info about display entities.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.bukkit.entity.TextDisplay;
+
+public class EntityOpacity extends EntityProperty<ElementTag> {
+
+    // <--[property]
+    // @object EntityTag
+    // @name opacity
+    // @input ElementTag(Number)
+    // @description
+    // A text display entity's text opacity, from 0 to 255.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java:15]: Should include notes about the odd behavior near-zero you observed in the meta.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public void setPropertyValue(ElementTag value, Mechanism mechanism) {
+        if (mechanism.requireInteger()) {
+            int opacity = value.asInt();
+            if (opacity < 0 || opacity > 255) {
+                mechanism.echoError("Invalid opacity specified, must be between 0 and 255.");
+                return;
+            }
+            as(TextDisplay.class).setTextOpacity((byte) opacity);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java:42]: I notice you convert from signed to unsigned in `get`, but not in `set` - is this intentional? Does this actually behave as expected?

Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java:42]: I'm pretty sure this is right? Correct me if I'm wrong, but according to my testing (and google) this should convert it properly (and it does look right in-game, other then the 0-3 edge case):
```
[19:34:34 INFO]: +> [EntityOpacity] Setting opacity: 255
[19:34:34 INFO]: +> [EntityOpacity] Raw opacity: -1

[19:34:40 INFO]: +> [EntityOpacity] Setting opacity: 128
[19:34:40 INFO]: +> [EntityOpacity] Raw opacity: -128

[19:34:43 INFO]: +> [EntityOpacity] Setting opacity: 127
[19:34:43 INFO]: +> [EntityOpacity] Raw opacity: 127

[19:34:58 INFO]: +> [EntityOpacity] Setting opacity: 234
[19:34:58 INFO]: +> [EntityOpacity] Raw opacity: -22
```

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityDisplay.java (hunk 3 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityDisplay.java:19]: Added synonyms for both

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java (hunk 5 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityOpacity.java:15]: Added a note for that

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityInterpolationDuration.java (hunk 4 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityInterpolationDuration.java:15]: Added an interpolation specific meta and linked it for now, let me know if it should be expanded

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityTranslation.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    // Can be interpolated, see <@link language Display entity interpolation>.
+    // -->
+
+    public static boolean describes(EntityTag entity) {
+        return entity.getBukkitEntity() instanceof Display;
+    }
+
+    @Override
+    public LocationTag getPropertyValue() {
+        Vector3f scale = as(Display.class).getTransformation().getTranslation();
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityTranslation.java]: `scale`?

Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityTranslation.java]: Oop, copypasta - fixed

**Pull request merged**